### ISSUE TITLE
Publish the SHA image as well in Quay when updating main and latest tags

### ISSUE THREAD

**Issue by nluaces** [skupperproject/skupper-router#292]
Could it be possible to change the build workflow and publish the version associated with the hash SHA in Quay?
That way previous versions of skupper-router would be still accessible for testing purposes.

What do you think?

**Comment by jiridanek** [comment]
Sounds good to me!

Just one question: what will be cleaning up old images? One image is few hundred MB, not sure how much compression and sharing of unchanged layers helps, but still, the data usage will go up with every commit, IIUC.

Does quay.io have any limits?

**Comment by ganeshmurthy** [comment]
If I understand correctly, you want the commit sha as the quay.io image tag ?

**Comment by nluaces** [comment]
@ganeshmurthy that's correct; but if you see any concerns and prefer to establish another type of tag that is also fine, because what @jiridanek mentioned is quite important; it would require cleaning up old images.

**Comment by ganeshmurthy** [comment]
You could create a branch from a specific commit sha and in that branch change the version in the VERSION.txt to that commit sha and push this branch up to your git repo. Then go to *your* Github URL - https://github.com/ganeshmurthy/skupper-router/actions/workflows/release.yml (this is my URL) and then click on Run Workflow and select the branch. The release.yaml will tag the image using this - https://github.com/ganeshmurthy/skupper-router/blob/main/.github/workflows/release.yml#L35

Please let me know if that works for you

**Comment by jiridanek** [comment]
> That way previous versions of skupper-router would be still accessible for testing purposes.

For how long should the previous versions remain accessible?

What about this alternative scheme: You'd be using the `latest` tag in skupper CI, same way as now, but every time the CI competelly passes, you'd add a `tested` tag to the skupper-router image used in the test. That way, you'd always have a guaranteed-working router image available, under the tested tag, without manual effort. Problem is, there will be only one tested image, not a longer history.

I have a few further thoughts on this:

## Untagged manifests

When a new image is pushed under a tag, the previous image that is replaced becomes untagged. Some repositories handle "untagged manifests" and user can configure retention policy. Here it is for Azure, https://docs.microsoft.com/en-us/azure/container-registry/container-registry-image-tag-version#delete-untagged-manifests. I did not find an setting like this for Quay.io, but it almost certainly is a thing, because it is possible to revert the tag updates in the WebUI.

Makes me wonder if the old versions are available for some time though the sha256 image hash from quay.

![image](https://user-images.githubusercontent.com/442720/162159866-bfe3d9fc-a8c7-42b2-a9ec-83976ab61865.png)

If yes, then you could somehow keep track of the hashes (or query quay.io for these older untagged images) and use them until they expirte.

## Nightly builds

Tagging images with sha256 of git commit of the sources makes sense, but it is somewhat impractical for humans to work with. Tagging by date, with having one build per day, might be more practical. The skupper-router git revision can be tracked as a LABEL on the image.

## Deploy a do-it-yourself image purging script

Doing the purging on our own as part of the build GitHub Actions seems doable, maybe the script can even be copied from somewhere, ready made.

## Quay image expiry

Quay has some sort of image expiry. The expiration can be set by building image with a LABEL. https://support.coreos.com/hc/en-us/articles/115001384693-Tag-Expiration

I dislike the LABEL approach, because the `:latest` and `:<long-hash>` images should be the same image, at the time new image is pushed. Adding LABEL modifies the image.

Expiry can be set manually in the UI. Not good. I did not find an API for it, or a cli tool to set it, but I am still hoping it is out there, somewhere.

## Why is this so hard to do?!? Using Quay.io for CI temp artifacts should be easy!

https://idbs-engineering.com/containers/2019/08/27/auto-expiry-quayio-tags.html

**Comment by nluaces** [comment]
@jiridanek thanks for your input, now I think that the option of having only one previous tag called `:tested` with the version that was located in `:latest` before initiating the build would be enough. I agree with you, it is absolutely more practical using that label that using a hash. Also, that way the problem of generating so many tags would be solved.

> Makes me wonder if the old versions are available for some time though the sha256 image hash from quay.

I tried pulling the image using the sha256 image hash, but I couldn't (I believe that happened because of the absence of manifest).

@ganeshmurthy thank you, the solution that you have provided works for me, I think this issue can be closed.

**Comment by jiridanek** [comment]
> > Makes me wonder if the old versions are available for some time though the sha256 image hash from quay.
> 
> I tried pulling the image using the sha256 image hash, but I couldn't (I believe that happened because of the absence of manifest).

I tried that too and it is not working with quay. I think it would work with Azure, according to their docs I linked, but we don't have any images there, so even if it works, it does not help.
